refl1d/probe.py

```diff
@@ -1561,12 +1565,44 @@ def measurement_union(xs):
     """
     Determine the unique (T, dT, L, dL) across all datasets.
     """
-
     # First gather a set of unique tuples in angle and wavelength
     TL = set()
     for x in xs:
         if x is not None:
-            TL |= set(zip(x.T+x.theta_offset.value, x.dT, x.L, x.dL))
+            TO = x.theta_offset.value
+            TO_lower, TO_upper = (TO, TO) if x.theta_offset.fixed else x.theta_offset.bounds.limits
+            # Sort by T
+            idxT = np.argsort(x.T)
+            T = x.T[idxT]
+            dT = x.dT[idxT]
+            L = x.L[idxT]
+            dL = x.dL[idxT]
+
+            if np.isinf(TO_upper) or np.isinf(TO_lower):
+                print(x.theta_offset.bounds, x.theta_offset.bounds.limits)
+                raise ValueError("cannot extend theta range to infinity from theta_offset")
+
+            # this assumes that negative theta values still add theta_offset normally
+            lower_step = dT[0] / 5.0
+            upper_step = dT[-1] / 5.0
+            T_lower_extension = np.arange(T[0] - TO_lower, T[0], lower_step, dtype=T.dtype)
+            dT_lower_extension = np.ones_like(T_lower_extension, dtype=dT.dtype) * dT[0]
+            L_lower_extension = np.ones_like(T_lower_extension, dtype=L.dtype) * L[0]
+            dL_lower_extension = np.ones_like(T_lower_extension, dtype=dL.dtype) * dL[0]
+
+            # start at upper end and add steps backward... then reverse.
+            T_upper_extension = np.arange(T[-1] + TO_upper, T[-1], -upper_step, dtype=T.dtype)[::-1]
+            dT_upper_extension = np.ones_like(T_upper_extension, dtype=dT.dtype) * dT[-1]
+            L_upper_extension = np.ones_like(T_upper_extension, dtype=L.dtype) * L[-1]
+            dL_upper_extension = np.ones_like(T_upper_extension, dtype=dL.dtype) * dL[-1]
+
+            T = np.hstack([T_lower_extension, T, T_upper_extension])
+            dT = np.hstack([dT_lower_extension, dT, dT_upper_extension])
+            L = np.hstack([L_lower_extension, L, L_upper_extension])
+            dL = np.hstack([dL_lower_extension, dL, dL_upper_extension])
+
+            TL |= set(zip(T, dT, L, dL))
```

Comment:
Should we alter the __init__ for PolarizedNeutronProbe so that it checks for a fixed geometry across all cross sections, then store only the one?

Reply:
If we oversample and extend in Q-space, we will not be able to support PolarizedNeutronProbe.unique_L and PolarizedNeutronProbe._L_idx attributes, which are calculated from an explicitly oversampled L, and there is no unique way to go back from Q -> T,L  ... so supporting wavelength-dependent SLD will get hard, will it not?

Reply:
I guess the code is already not oversampling in L properly, anyway, since we are currently sorting on T and then oversampling L in T order.

Reply:
I think we need to create a θ-λ grid for the energy dependent case, with separate oversampling in θ and λ. Render the microslabs and calculate the reflectivity for each λ separately, then integrate over θ-λ patches for the resolution.

I had originally planned to allow a separate ρ for each q, but hadn't worked out the details of how to generate them.

Anyway, we could drop the L_idx attribute since it was never used.